### PR TITLE
Replace $http with fetch in OAuthClient

### PR DESCRIPTION
**Pull request by robertknight**
This PR replaces Angular's `$http` service with [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) for calls to h's OAuth token endpoints.

This is part of #974.

The first commit adds a polyfill for the fetch API for older browsers and improves the documentation on how to add new polyfills in future.

The second commit modifies the `OAuthClient` utility class to use `window.fetch` instead of the `$http` service and updates the tests to use the fetch-mock package instead of a manual mock of Angular's `$http` service.

----

**To test this branch manually:**

- Test that log in and log out works
- Open the browser dev tools and go to the Network tab
- Test log in / log out of the client again. On master, you'll see requests made to the "token" and "revoke" endpoints using xhr. On this branch, the devtools should indicate they were made via "fetch" instead.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -121,11 +121,21 @@ const appBundles = [{
 // based on major ECMAScript version or DOM API. Some large polyfills
 // (eg. for String.prototype.normalize) are additionally separated out into
 // their own bundles.
+//
+// To add a new polyfill:
```
Comment (lyzadanger) [inline gulpfile.js]: Thank you for these comments!

File: src/boot/boot.js [hunk 2]
```diff
@@ -96,10 +96,15 @@ function bootHypothesisClient(doc, config) {
  */
 function bootSidebarApp(doc, config) {
   const polyfills = polyfillBundles([
+    // JS polyfills.
     'es2015',
     'es2016',
     'es2017',
     'string.prototype.normalize',
+
+    // DOM polyfills. These are loaded after the JS polyfills as they may
```
Comment (lyzadanger) [inline src/boot/boot.js]: Comments FTW

File: src/sidebar/util/oauth-client.js [hunk 3]
```diff
@@ -15,22 +15,23 @@ const random = require('./random');
 
 /**
  * Return a new TokenInfo object from the given tokenUrl endpoint response.
- * @param {Object} response - The HTTP response from a POST to the tokenUrl
- *                            endpoint (an Angular $http response object).
- * @returns {TokenInfo}
+ * @param {Response} response - The HTTP response from a POST to the tokenUrl
```
Comment (lyzadanger) [inline src/sidebar/util/oauth-client.js]: This is great...real `Response`s!

File: src/sidebar/util/oauth-client.js [hunk 4]
```diff
@@ -208,11 +206,19 @@ class OAuthClient {
    * @param {Object} data - Parameter dictionary
    */
   _formPost(url, data) {
```
Comment (lyzadanger) [inline src/sidebar/util/oauth-client.js]: Does this method have more general usefulness outside of OAuth?

Comment (robertknight) [reply src/sidebar/util/oauth-client.js]: We don't post form data anywhere else in the client at the moment. If we did, then it might be useful.

**Pull request merged**